File: Leetcode/Problems C++/189. Rotate Array.cpp
class Solution {
public:
    void reverse(vector<int>&nums,int beg,int end)
    {
        end = end-1;
        while(beg<end)
        {
            int temp = nums[beg];
            nums[beg] = nums[end];
            nums[end] =temp;
            beg++;
            end--;
        }
    }
    void rotate(vector<int>& nums, int k) {
        int size = nums.size();
        k=k%size;
        if(size<=1 || k==0)
            return;
        
        reverse(nums,0,size);
        reverse(nums,0,k);
        reverse(nums,k,size);
    }
};

/*
Approach 1 : move 1 by 1 for k times. O(n)*k == O(nk)
Approach 2 : extra space. O(n) memory,O(n) space
Approach 3 : reversing O(n)
*/

File: Leetcode/Problems Java/189. Rotate Array.java
class Solution {
    // public void rotate(int[] nums, int k) {
    //     int n = nums.length;
    //     k %= n;
    //     if(k==0 || (k==1 && n == 1))return;
    //     while(k>0){
    //         k--;
    //         int temp = nums[n-1];
    //         for(int i=n-1;i>=1;i--)
    //             nums[i] = nums[i-1];
    //         nums[0] = temp;
    //     }
    // }
    //TLE O(n*k)

    private void reverse(int[]nums, int l, int r){
        while(l<r){
            int t = nums[l];
            nums[l++] = nums[r];
            nums[r--] = t;
        }
    }
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k %= n;
        if(k==0 || (k==1 && n == 1))return;
        reverse(nums, 0, n-k-1);
        reverse(nums, n-k, n-1);
        reverse(nums, 0, n-1);
    }
}
